src/rest_api/classes/laboratory/widgets/members.clj

```diff
@@ -1,31 +1,108 @@
 (ns rest-api.classes.laboratory.widgets.members
   (:require
-    [rest-api.classes.generic-fields :as generic]
+    [datomic.api :as d]
+    [clojure.string :as str]
+    [pseudoace.utils :as pace-utils]
+    [rest-api.formatters.date :as date]
     [rest-api.formatters.object :as obj :refer [pack-obj]]))
 
-(defn- get-lineage-data [m]
- ;; supervised-by supervised worked-with
- (let []
-  {:name (pack-obj m)
-   :k (keys m)
-   :d (:db/id m)
-   :level nil
-   :start_date nil
-   :end_date nil
-   :duration nil
-   }
-  ))
+(defn- generate-map [oid]
+  (let [other-person-id (get oid 0)
+        other-person-name (get oid 1)
+        json-data (get oid 2)
+        roles (filter #(re-find #"role" (str %)) (keys json-data))
+        level
+        (for [role roles
+              :let
+              [level (-> role
+                         (str/replace #":role/" "")
+                         (str/replace #"-" " ")
+                         (str/capitalize))]] level)
+        from
+        (for [role roles
+              :let [rfrom (keyword (-> role (str/replace #"/" ".") (str/replace #":" "")) "from")
+                    from (if (contains? (first (json-data role)) rfrom)
+                           (date/format-date5 ((first (json-data role)) rfrom)))]] from)
+        to
+        (for [role roles
+              :let [ rto (keyword (-> role (str/replace #"/" ".") (str/replace #":" "")) "to")
+                    to (if (contains? (first (json-data role)) rto)
+                         (date/format-date5 ((first (json-data role)) rto)))]] to)
+        duration
+        (for [role roles
+              :let [rfrom (keyword (-> role (str/replace #"/" ".") (str/replace #":" "")) "from")
+                    rto (keyword (-> role (str/replace #"/" ".") (str/replace #":" "")) "to")
+                    from (if (contains? (first (json-data role)) rfrom)
+                           (date/format-date5 ((first (json-data role)) rfrom)))
+                    to (if (contains? (first (json-data role)) rto)
+                         (date/format-date5 ((first (json-data role)) rto)))
+                    duration (str from " - " to)]] duration)]
 
-(defn current-members [l]
-  {:data (some->> (:laboratory/registered-lab-members l)
-                  (map get-lineage-data))
-   :description "current members of the laboratory"})
+    { (keyword other-person-id) {:level (first level)
+                                 :person other-person-id
+                                 :duration (first duration)
+                                 :start_date (first from)
+                                 :end_date (first to) } }))
 
-(defn former-members [l]
-  {:data (some->> (:labloratory/past-lab-members l)
-                  (map get-lineage-data))
-   :description "former members of the laboratory"})
+
+(defn- generate-lineage [laboratory]
+  (let [db (d/entity-db laboratory)
+        reps (:laboratory/representative laboratory)
+        lineage
```

Comment:
I think that your code would simplify a lot if you were to remove extra variables like "lineage" here. Instead of defining lineage in the let an then returning lineage inside the let. The logic for creating lineage should just be in side the let. After my next suggestion you should be able to get rid of the let all together. But it would be good to take out these extra variables. Once they are removed I think i could give much better suggestions on how to simplify the code.

Reply:
I'm not sure how you mean, isn't the logic for lineage inside the (let) already ?  I wanted to pass 'registered-lab-members' and 'past-lab-members' as variables for the query, so that former-members and current-members could be calling the same function passing either string, then the function defining lineage could be written once inside that new function, but I couldn't figure out the syntax for that, and got this to work.